File: Leetcode/Container With Most Water.py
'''
Input: height = [1,8,6,2,5,4,8,3,7]
Output: 49
'''
class Solution:
    def max2(self, height: list[int])->int:
        '''
        start, k, max1 = 0, 0, 0
        for j in range(0, len(height)):
            start = height[j]
            for i in range(j + 1, len(height)):
                indexDif = (i - k) - j
                max1 = max(max1,(min(start, height[i]) * indexDif))
        return max1
        '''
        # Another way to solution
        start, end, maxi =0, len(height)-1,0
        while start < end:
            maxi = max(maxi, (min(height[start],height[end])*(end-start)))

            if height[start]<height[end]:
                start= start+1
            else: end= end-1
        return maxi

solution = Solution()
print(solution.max2([1,7,9,2,5,4,8,9,6]))

File: Leetcode/Container With Most Water.java
/*
Input: height = [4,3,2,1,4]
Output: 16
Input: height = [1,8,6,2,5,4,8,3,7]
Output: 49
*/

public class ContainersWithMostWater {

    public int maxArea(int[] height) {
        int start = 0, end = height.length - 1;
        int maxi = 0;
        while(start < end){
            int mini = Math.min(height[start],height[end]);
            maxi = Math.max(maxi, (end - start)*mini);
            if(height[start] < height[end])start++;
            else end--;
        }
        return maxi;
    }

    public static void main(String args[]){
        ContainersWithMostWater containers = new ContainersWithMostWater();
        System.out.println(containers.maxArea(new int[]{4,3,2,1,4}));
    }
}
